ExpressLRS/ExpressLRS
# Uber TX

# Problem
OMG! So many targets and firmwares!

# Solution
Uber TX!

This PR is based on top of the binary firmware PR, the reason for that is that it helps consolidate the STM based targets as well.

This is a single unified firmware for ESP32 TX targets that loads a hardware descriptor (`hardware.json`) from the SPIFFS partition on startup.
If the `hardware.json` file is not found the module boots straight to wifi mode and you can go to unlock-higher-power.
There is a drop target for dropping pre-defined hardware files or you can manually enter the values for your specific target.

Also, the user-defines are now included in a SPIFFS file and can be changed/setup by going to http://elrs-tx.local/options.html. The binding phrase to UID script is included on the web-page for simplicity as well.
This file is a json file with the following entries
- `uid`
- `wifi-on-interval` (in seconds) (defaults to 30)
- `wifi-ssid`
- `wifi-password`
- `tlm-interval` (in milliseconds) (defaults to 320)
- `fan-runtime` (in seconds) (defaults to 30)
- `no-sync-on-arm` (defaults to false)
- `uart-inverted` (default to true)
- `unlock-higher-power` (defaults to false)

Also, the "device name" as sent to the LUA script is also stored in a SPIFFS file called `device.ini`. This is a file with a single line with the text string for the device name.

The 3 files can be directly downloaded from the SPIFFS partition by going to the appropriate URL e.g. http://elrs-tx.local/hardware.json. Extra functionality has been added so these files can be uploaded using HTTP PUT to the same URL. Also added a `/reboot` which reboots the device after you've uploaded the 3 control files (useful for Configurator).

This reduces the number of ESP32 to targets from 26 to 2 (one for SX1280 and one for SX127x)

This also solves the problem of the proliferation of many user-defined DIY targets where there are only slight differences.
Now a user can easily create their own target and keep a copy of their hardware file and upload it again if they do a full chip erase.

## UPDATE
I've added the ability to attach the json files to the end of the firmware and adjusted the build process to do this automatically. Just put the desired .json file in the data directory as `hardware.json` and do a build. NOT a build and flash. Then flash separately. The end of the build phase will attach the device_name, generated `options.json` and `hardware.json` to the of the binary. So not more uploading to SPIFFS.

# How to test 1
1. Copy a hardware configuration file from the `hardware` folder to the data folder and name the file `hardware.json`
2. Build, not "Build and flash" `UBER_2400_TX_via_UART` (or other upload method)
3. Flash the firmware file
4. the TX module will reboot and should be running as normal
5. 🎉 

# How to test 2
1. Use build and flash `UBER_2400_TX_via_UART`
2. goto the wifi page http://elrs_tx.local and you should see the target configuration page
3. drag and drop one of the target files from the `hardware` directory and press the `Save target configuration` button at the bottom of the page.
4. the TX module will reboot and should be running as normal
6. 🎉 

# What has been tested
- Axis Thor
- BETAFPV 900 Micro
- BETAFPV 2400 Micro
- BETAFPV 2400 Micro 1W
- DIY E28
- DIY TTGO 900 V2
- HappyModel ES24 Pro
- iFlight 2400
- iFlight 900
- Jumper AION Nano
- Jumper T-Pro
- Namimno Flash OLED
- Radiomaster Zorro
- Vantac Lite

# What's Left
1. Uber RX!
